ci(nix): Migrate to crane (and speed up Nix CI)

9glenda opened the pull request:
Closes #411 
Closes  #464

# TODO
- [ ] Fix failing tests
- [x] clippy
- [x] trycmd
- ~~Cross compilation. https://crane.dev/examples/cross-rust-overlay.html~~ should be done in a extra PR
- ~~Maybe optional musl cross compilation~~
- [x] add clippy to packages
- [x] add `zip`
- [x] use flake parts
- [x] Check for name changes in outputs breaking the Justfile
- [x] apply patch from cafkafk for the Justfile.
- [x] bring back Pre-commit-hooks checks
- [x] rime (error 520?)
- [x] treefmt packages refactor
- [x] dev shell
## Changes interesting for review

- eza.meta is generated from cargo toml in favor of hard coding it
- version changed from latest to "${version}-git"
- cargo plugins used in `Justfile` added to dev shell

## Notes (links for me to keep in mind)
- [working cross compilation](https://github.com/DeterminateSystems/flake-checker/blob/main/crane.nix)

- [cafkafk's idea to speed up the build](https://github.com/Mic92/nix-fast-build) 

## NOTE
due to nix fmt only running `statix fix` statix lint's which don't have quickfixes will not be fixed by nix fmt and will throw an error at the pre-commit-hooks

9glenda committed:
ci(nix): speed up Nix CI

9glenda committed:
ci(nix): fix derivation path

9glenda committed:
ci(nix): only build on x86_64-linux

9glenda committed:
ci(nix): initial infrastructure for `crane`

the main advantage of [crane](github.com/ipetkov/crane) is that it is faster then naersk

9glenda committed:
ci(nix): add flake checks

Also the eza package now supports man and completions.
This makes it fully compatible with the old one.

9glenda committed:
style: fix clippy warnings

line src/options/view.rs:434 might be interesting for the review

9glenda committed:
ci(nix): added trycmd test and trydump package

9glenda committed:
ci(nix): only run on code changes

9glenda committed:
ci(nix): moved clippy to packages

cafkafk reviewed (commented):
This is cool, thanks for doing this :heart: 
---

I wonder what's causing this:

![2023-10-03_11-50](https://github.com/eza-community/eza/assets/89321978/a7f3870b-d11b-43f9-a97b-de327670fef2)
---

The new test runner mostly looks good, but wonder if we can fix this output being slightly broken. Honestly I'm fine with it being a little broken over not having colors thou.
![2023-10-03_11-51](https://github.com/eza-community/eza/assets/89321978/a0eb137f-d5e4-485d-9f2a-ce72d05837ab)

Also, why are those tests skipped?

cafkafk commented:
> @cafkafk I modified some of the apparently skipped tests to fail. This also mad the CI fail. I assume the skipped is just a visual bug. when installing the cargo-nextest plugin and running the command manually not through nix there's no skipped text. ![image](https://user-images.githubusercontent.com/69043370/272243254-5c8a4b17-f38a-4b05-a4a4-b13ab944e80a.png)

Hmm, if it's just a visual bug we can live with it for now. Where should we mention this bug so upstream can fix it?

cafkafk commented on flake.nix:
```diff
@@ -86,77 +100,84 @@
+        # For `nix build`
+        packages = {
+          default = eza;
+          trydump = craneLib.cargoNextest (commonArgs
+            // {
+              # include test files
+              src = ./.;
+              inherit cargoArtifacts;
+              mode = "test";
+              doCheck = true;
+              # No reason to wait for release build
+              release = false;
+              # buildPhase files differ between dep and main phase
+              singleStep = true;
+              # set itests files creation date to unix epoch
+              buildPhase = ''touch --date=@0 tests/itest/*; rm tests/cmd/*.stdout || echo; rm tests/cmd/*.stderr || echo;'';
+              cargoTestOptions = ["--features nix"];
+              TRYCMD = "dump";
+              postInstall = ''
+                cp dump $out -r
+              '';
+            });
```
This doesn't seem to generate the nix tests anymore in `results/dump`

cafkafk commented:
I also think we should change this in the `Justfile`, seems like the current glob catches `unix`

```
diff --git a/Justfile b/Justfile
index d50984f6..e048b95a 100644
--- a/Justfile
+++ b/Justfile
@@ -265,8 +265,8 @@ gen_test_dir:
 #
 # WARNING: this can cause loss of work
 @idump:
-    rm ./tests/cmd/*_nix.stderr -f || echo  
-    rm ./tests/cmd/*_nix.stdout -f || echo
+    rm ./tests/cmd/*nix.stderr -f || echo  
+    rm ./tests/cmd/*nix.stdout -f || echo
     nix build -L ./#trydump
-    cp ./result/dump/*_nix.* ./tests/cmd/
+    cp ./result/dump/*nix.* ./tests/cmd/
```

(I can't push to this PR for some reason so here's a patch)

cafkafk commented:
Sorry for making your life harder, but I changed the flake a bit in #476 >_<

cafkafk commented:
I realized today that we should probably add `zip` to the devshell, the justfile uses it for the gh-release recipe

9glenda committed:
ci(nix): added flake-compat

[Flake-compat](https://github.com/edolstra/flake-compat) allows the flake to be usable without flakes through legacy nix.

9glenda committed:
Merge branch 'main' into main

Signed-off-by: 9glenda <plan9git@proton.me>

cafkafk commented:
> @cafkafk I added [flake-compat](https://github.com/edolstra/flake-compat) last commit. I think there's nothing wrong with it since it allows non flake users to also use the flake through legacy nix. If you have any issues with flake-compat we can also remove it.

No I have no problem with adding it, the more can use it the better IMO!

9glenda committed:
build(deb): switched `extended-description` from flake

Changed the `extended-description` to the `longDescription` used in the flake.
This is related to the `flake.nix` refactor because of wanting to generate `package.meta` from `Cargo.toml`

9glenda committed:
ci(nix): initial infra structure for `flake-parts`

9glenda committed:
Merge branch 'main' into main

Signed-off-by: 9glenda <plan9git@proton.me>

9glenda committed:
style: removed comments in flake.nix

9glenda committed:
ci(Justfile): Justfile patch from @cafkafk

```patch
diff --git a/Justfile b/Justfile
index e048b95a..d50984f6 100644
--- a/Justfile
+++ b/Justfile
@@ -265,8 +265,8 @@ gen_test_dir:
 #
 # WARNING: this can cause loss of work
 @idump:
-    rm ./tests/cmd/*nix.stderr -f || echo
-    rm ./tests/cmd/*nix.stdout -f || echo
+    rm ./tests/cmd/*_nix.stderr -f || echo
+    rm ./tests/cmd/*_nix.stdout -f || echo
     nix build -L ./#trydump
-    cp ./result/dump/*nix.* ./tests/cmd/
+    cp ./result/dump/*_nix.* ./tests/cmd/
```

9glenda committed:
ci(nix): added cargo plugins used in `Justfile` to devShell

9glenda committed:
Merge branch 'main' into main

Signed-off-by: 9glenda <plan9git@proton.me>

9glenda committed:
style: fmt

9glenda committed:
Merge branch 'main' into main

Signed-off-by: 9glenda <plan9git@proton.me>

9glenda committed:
ci(nix): Merge branch 'eza/main' into '9glenda/main'

Signed-off-by: 9glenda <plan9git@proton.me>

ci(nix): fix checks

9glenda committed:
ci(nix): merge 'eza/main' into '9glenda/main'

Signed-off-by: 9glenda <plan9git@proton.me>

ci(nix): merge 'eza/main' into '9glenda/main'

9glenda committed:
ci(nix): add trycmd and trydump

ci(nix): add trycmd to tests

9glenda committed:
build: update Justfile

9glenda committed:
ci(nix): remove unused flake inputs

ci(nix): disable color for nextest

9glenda committed:
ci(nix): refactor treefmt

9glenda committed:
ci(nix): fix #464

9glenda committed:
docs: fix typo in flake path

MartinFillon commented:
Just FYI #592 is coming

MartinFillon commented:
well it is you think it is better as I aim to delete _regular_ nix itest and move them to generated dir with time. But now that I think about it my pr just separating both doesnt makes sense, I shall just migrate everything in one batch.

9glenda committed:
ci: Merge branch 'eza/main' into 9glenda/main

Signed-off-by: 9glenda <plan9git@proton.me>

cafkafk reviewed (commented):
This isn't an issue, but I had to upgrade nix `v2.17.1 -> v2.18.1` for this to work, but that's fine imo.

I've not had enough time to review this yet but will do a full review asap

cafkafk reviewed (changes requested):
I'm not sure why (except for one version test), but the integration tests fail. I'll try to regenerate them and see if it works.

cafkafk commented:
> I'm not sure why (except for one version test), but the integration tests fail. I'll try to regenerate them and see if it works.

There are still at least two failures, I'm not sure why they fail thou >_<

MartinFillon commented:
the tests run in ci are trycmd and trycmd-local also calls ptests so that ci is not too long @9glenda

cafkafk commented:
> > I'm not sure why (except for one version test), but the integration tests fail. I'll try to regenerate them and see if it works.
> 
> @cafkafk Are the integration tests just trycmd or also trycmd-local? 

IIRC it *should* be just trycmd, and trycmd-local *should* be reserved for being run locally, but my memory may fail me.

> The trycmd-local tests complain about sudo not being found but that's also the case on the main branch iirc.

Yes, it shouldn't be a blocking problem I think?

cafkafk reviewed (changes requested):
There are also some merge conflicts that will need to be solved, I'm likely gonna do a lot of merges today, so may wanna wait trying to solve it until after today.
